man/folder.man
.TH FOLDER %manext1% 2016-08-29 "%nmhversion%"
.
.\" %nmhwarning%
.
.SH NAME
folder, folders \- set/list current nmh folder/message
.SH SYNOPSIS
.HP 5
.na
.B folder
.RB [ \-help ]
.RB [ \-version ]
.RI [ +folder ]
.RI [ msg ]
.RB [ \-all " | " \-noall ]
.RB [ \-create " | " \-nocreate ]
.RB [ \-fast " | " \-nofast ]
.RB [ \-header " | " \-noheader ]
.RB [ \-recurse " | " \-norecurse ]
.RB [ \-total " | " \-nototal ]
.RB [ \-list " | " \-nolist ]
.RB [ \-push " | " \-pop ]
.RB [ \-pack " | " \-nopack ]
.RB [ \-print " | " \-noprint ]
.RB [ \-verbose " | " \-noverbose ]
.HP 5
.B folders
is equivalent to
.B folder
.B \-all
.ad
.SH DESCRIPTION
When
.B folder
is given the
.B \-print
switch (the default), it lists: the current folder, the number
of messages in it and their range (low-high), the folder's
current message, and an indication of extra files, if any.
For example:
.PP
.RS 5
.nf
.ta \w'/rnd/phyl/Mail/EP 'u +\w'has ddd messages 'u +\w'(ddd\-ddd); 'u
inbox+	has \016 messages	(\0\03\-\022);	cur=\0\05.
.fi
.RE
.PP
If a
.I +folder
and/or
.I msg
are specified, they will become the current folder and/or message.
(When a
.I +folder
argument is given, this corresponds to a \*(lqcd\*(rq operation
in the shell; when no
.I +folder
argument is given, this corresponds roughly to a \*(lqpwd\*(rq
operation in the shell.) If
.B \-noprint
is specified, a
.I +folder
and/or
.I msg
may still be specified to set the current folder and/or message,
but the folder summary will not be printed.
.PP
If the specified (or default) folder doesn't exist,
.B folder
will ask the user if the folder should be created; when
standard input is not a tty, the answer is assumed to be
\*(lqyes\*(rq.
The
.B \-create
switch tells
.B folder
to create new folders without query.
(This is an easy way to create an empty folder for later use.)
Specifying
.B \-nocreate
will cause
.B folder
to print an error message and exit without creating a folder.
.SS "Multiple Folders"
Specifying
.B \-all
will produce a summary line for each top-level folder in the user's
.B nmh
directory, sorted alphabetically.  (If
.B folder
is invoked by a name ending with \*(lqs\*(rq (e.g.
.BR folders ),
.B \-all
is assumed.)  Specifying
.B \-recurse
with
.B \-all
will include a line for all sub-folders.
These folders are all preceded by any read-only folders, which occur as
.RI \*(lq atr\-cur\- \*(rq
entries in the user's
.B nmh
context.  For example:
.PP
.RS 5
.nf
.ta \w'/rnd/phyl/Mail/EP 'u +\w'has ddd messages 'u +\w'(ddd\-ddd); 'u
FOLDER	\0\0\0\0\0\0# MESSAGES	RANGE	CUR     (OTHERS)
/var/work/folder	has \035 messages	(\01\-\035);	cur=23.
/usr/bugs/Mail	has \082 messages	(\01\-108);	cur=82.
ff	has \0no messages.
inbox+	has \016 messages	(\03\-\022);	cur=\05.
mh	has \076 messages	(15\-\076);	cur=70.
notes	has \0\02 messages	(\01\-\0\02);	cur=\01.
ucom	has 124 messages	(\01\-124);	cur=\06; (others).
.ta \w'/rnd/phyl/Mail/EP has 'u
TOTAL = 339 messages in 7 folders
.fi
.RE
.PP
The \*(lq+\*(rq after
.I inbox
indicates that it is the current folder.
The \*(lq(others)\*(rq indicates that the folder
.I ucom
contains files which aren't messages;
these may either be sub-folders, or files that
don't belong under the
.B nmh
file naming scheme.
.PP
The header is output if either the
.B \-all
or
.B \-header
switches are specified.  It is suppressed by
.BR \-noheader .
.PP
The folder and message totals are output if either the
.B \-all
or
.B \-total
switches are specified.  They are suppressed by
.BR \-nototal .
.PP
If
.B \-fast
is given, only the folder name (or names, in the case of
.BR \-all )
will be listed.
(This is faster because the folders need not be read.)
.PP
If a
.I +folder
is given with the
.B \-all
switch,
.B folder
will, in addition to setting the current folder, list the top-level
sub-folders for the current folder (with
.BR \-norecurse )
or list all sub-folders under the current folder recursively (with
.BR \-recurse ).
.PP
If
.I msg
is supplied, together with
.I +folder
or without
.BR \-all ,
it will become the current message of
.I +folder
(if it had been supplied)
or the current folder.
.PP
The
.B \-recurse
switch lists each folder recursively, so its use effectively
defeats the speed enhancement of
.BR \-fast ,
since each folder must be searched for sub-folders.
Nevertheless, the combination of these options is useful.
.SS "Compacting a Folder"
The
.B \-pack
switch will compress the message names in the designated folders,
removing holes in message numbering.  The
.B \-verbose
switch directs
.B folder
to tell the user the general actions it is taking to compress the folder.
.SS "The Folder Stack"
The
.B \-push
switch directs
.B folder
to push the current folder
onto the
.IR folder-stack ,
and make the
.I +folder
argument the current folder.  If
.I +folder
is not given, the current folder and the
top of the
.I folder-stack
are exchanged.  This corresponds to the
\*(lqpushd\*(rq operation in the shell.
.PP
The
.B \-pop
switch directs
.B folder
to discard the top of the
.IR folder-stack ,
after setting the current folder to that value.
No
.I +folder
argument is allowed.  This corresponds to the \*(lqpopd\*(rq
operation in the shell.  The
.B \-push
and
.B \-pop
switches are mutually exclusive: the last occurrence of either one
overrides any previous occurrence of the other.
The use of either of these switches also sets
.BR \-list ,
by default.
.PP
The
.B \-list
switch directs
.B folder
to list the
.IR folder-stack .
No
.I +folder
argument is allowed.  After a successful
.B \-push
or
.BR \-pop ,
the
.B \-list
action is taken, unless a
.B \-nolist
switch follows them on the command line.  This corresponds
to the \*(lqdirs\*(rq operation in the shell.  The
.BR \-push ,
.BR \-pop ,
and
.B \-list
switches disable the default
.B \-print
behavior.
.SH FILES
.TP 20
$HOME/.mh_profile
The user's profile.
.SH "PROFILE COMPONENTS"
.PD 0
.TP 20
Path:
To determine the user's nmh directory.
.TP
Current\-Folder:
To find the default current folder.
.TP
Folder\-Protect:
To set mode when creating a new folder.
.TP
Folder\-Stack:
To determine the folder stack.
.\" ^lsproc:~^Program to list the contents of a folder
.PD
.SH "SEE ALSO"
.IR mhpath (1),
.IR refile (1),
.IR rmf (1)
.SH DEFAULTS
.PD 0
.TP 20
+folder
The current folder.
.TP
msg
\*(lqnone\*(rq
.TP
\-nofast
.TP
\-noheader
.TP
\-nototal
.TP
\-nopack
.TP
\-norecurse
.TP
\-noverbose
.TP
\-print
is the default if none of
.BR -list ,
.BR \-push ,
or
.B \-pop
are specified.
.TP
\-list
is the default if
.BR \-push ,
or
.B \-pop
is specified.
.PD
.SH CONTEXT
If
.I +folder
and/or
.I msg
are given, they will become the current folder and/or message.
.SH BUGS
There is no way to restore the default behavior
(to ask the user whether to create a non-existent folder)
after
.B \-create
or
.B \-nocreate
is given.


NEWS
# NEWS - Release notes for nmh 1.7
#

Welcome to nmh, the new version of the classic MH mail handling system.
It's been over three years since the last release of nmh, and there have
been a number of significant changes since the last release.  Long-time
MH and nmh uses should read careful the NOTEABLE CHANGES section, as there
are some significant changes to nmh behavior.  Otherwise, please see the
README and INSTALL files for help on getting started with nmh.

For news of future releases, subscribe to the low-volume
    https://lists.nongnu.org/mailman/listinfo/nmh-announce

---------------
NOTABLE CHANGES
---------------

The largest notable changes in the 1.7 release are:

- Complete unification of network security support.  All network protocols
  (currently, POP and SMTP) have been refactored to use a common set of
  security routines.  This means all protocols support all SASL mechanisms
  (via the Cyrus-SASL library) and TLS.  TLS support has been strengthened
  to perform certificate name validation and to require TLS 1.1 as a
  minimum protocol.  Also, all protocols can make use of the OAuth2/XOAUTH
  SASL mechanism, which is supported by Gmail.
- send(1) now supports adding switches to post(8) based on the address or
  domain of the email address in the From: header; this more easily allows
  users to support multiple identities.
- A generic facility for passing arguments to filter programs in repl(1)
  by use of the -convertargs switch.
- Native support for the manipulation of iCalendar requests; see mhical(1)
  for more details.

------------
NEW FEATURES
------------

The following are new features for the 1.7 release of nmh:

- When building from source, configure will derive ${prefix} from an existing
  nmh installation if it finds one in your $PATH.
- Added welcome message when nmh detects that its version changed.
- The default locations for configuration files and support binaries
  have been changed.  Configuration files now install into ${sysconfdir}/nmh,
  and support binaries are placed in ${libexecdir}/nmh.  If you are upgrading
  an existing installation you should look for old configuration files in
  ${sysconfdir} and merge any local customizations into the new files in
  ${sysconfdir}/nmh, then remove the old files.  ${libdir} will also contain
  obsolete support programs that should be removed.
- All TLS connections now perform certificate validation (including hostname
  matching) by default; can be disabled on a per-application basis.
- post now defaults to port 587 on 'smtp' message submission.
- A value of 0 for the width switch of scan(1), inc(1), ap(1), dp(1),
  fmttest(1), and mhl(1) now means as many characters as the format
  engine can produce [Bug #15274].  That amount is limited by internal
  buffers.
- If a component has trailing whitespace, e.g., body:component="> ",
  mhl now trims that whitespace off when filtering blank text lines.
- An "rtrim" flag has been added to mhl to remove any trailing
  whitespace from filtered text lines.  A corresponding "nortrim" flag
  has also been added.
- Added getmymbox and getmyaddr mh-format(5) function escapes.
- New -[no]changecur, -fixtype, -decodetypes, and -[no]crlflinebreaks switches
  have been added to mhfixmsg(1).
- mhfixmsg now removes an extraneous trailing semicolon from header
  parameter lists.
- Added -convertargs switch to repl(1), to pass arguments to programs
  specified in the user's profile or mhn.defaults to convert message
  content.
- Added mhical(1), to display, reply to, and cancel iCalendar (RFC 5545)
  event requests.
- Added multiply mh-format(5) function.
- "mhparam bindir" prints the path to the directory containing the public
  executables (${bindir}).
- New "-prefer" switch for mhshow (and mhlist and mhshow), to allow specifying
  the preferred content types to show, if present in a multipart alternative.
- mh-format now has %(kilo) and %(kibi) functions, to allow printing
  numbers with SI or IEC quantities, e.g. "10K", "2.3Mi".
- Support for the -sendmail flag to send/post to change the sendmail
  binary when using the sendmail/pipe MTS.
- Added support to send(1) to specify switches to post(1) based on address or
  domain name in From: header line in message draft.
- post(8) -snoop now attempts to decode base64-encoded SMTP traffic.
- folder(1) -nocreate now prints a warning message for a non-existent folder.
- mhfixmsg(1) now allows -decodetext binary, though 8bit is still the default.
- inc(1) and msgchk(1) now support TLS encryption natively.
- All network protocols support the XOAUTH authentication mechanism.
- Support for SMTPUTF8 (RFC 6531) has been added.  mhshow(1) already supported
  RFC 6532, assuming all 8-bit message header field bodies are UTF-8 and use
  of a UTF-8 locale.
- mhfixmsg now replaces RFC 2047 encoding with RFC 2231 encoding of name and
  filename parameters in Content-Type and Content-Disposition headers,
  respectively.
- If a message body contains 8-bit bytes, post(8) uses SMTP 8BITMIME if the
  server supports it.  If not, post fails with a message to the user to
  encode the message for 7-bit transport.
- Fewer lseek(2)s will be used when reading headers in the common case.
- ./configure's --enable-debug has been removed;  it did nothing.
- configure now defaults to enabling each of TLS and Cyrus SASL if the
  necessary headers and libraries are found.
- Moved build_nmh to top-level directory.
- Better error reporting for connections to network services.

---------
BUG FIXES
---------

- The format scanner no longer subtracts 1 from the width.  This has the
  effect of no longer counting the trailing newline in the output of
  scan(1), inc(1), and the other programs that rely on the format scanner.
- The first character of some very short (less than 4 characters) message
  bodies is no longer dropped.
- Single-character headers can be reliably formatted, etc., instead of
  apparently being missing.

uip/folder.c
/* folder.c -- set/list the current message and/or folder
 *             -- push/pop a folder onto/from the folder stack
 *             -- list the folder stack
 *
 * This code is Copyright (c) 2002, 2008, by the authors of nmh.  See the
 * COPYRIGHT file in the root directory of the nmh distribution for
 * complete copyright information.
 */

#include "h/mh.h"
#include "sbr/getarguments.h"
#include "sbr/concat.h"
#include "sbr/seq_setprev.h"
#include "sbr/seq_setcur.h"
#include "sbr/seq_save.h"
#include "sbr/smatch.h"
#include "sbr/ssequal.h"
#include "sbr/getcpy.h"
#include "sbr/m_convert.h"
#include "sbr/getfolder.h"
#include "sbr/folder_read.h"
#include "sbr/folder_pack.h"
#include "sbr/folder_free.h"
#include "sbr/context_save.h"
#include "sbr/context_replace.h"
#include "sbr/context_del.h"
#include "sbr/context_find.h"
#include "sbr/brkstring.h"
#include "sbr/ambigsw.h"
#include "sbr/path.h"
#include "sbr/print_version.h"
#include "sbr/print_help.h"
#include "sbr/error.h"
#include "sbr/crawl_folders.h"
#include "h/done.h"
#include "h/utils.h"
#include "sbr/m_maildir.h"

#define FOLDER_SWITCHES \
    X("all", 0, ALLSW) \
    X("noall", 0, NALLSW) \
    X("create", 0, CREATSW) \
    X("nocreate", 0, NCREATSW) \
    X("fast", 0, FASTSW) \
    X("nofast", 0, NFASTSW) \
    X("header", 0, HDRSW) \
    X("noheader", 0, NHDRSW) \
    X("pack", 0, PACKSW) \
    X("nopack", 0, NPACKSW) \
    X("verbose", 0, VERBSW) \
    X("noverbose", 0, NVERBSW) \
    X("recurse", 0, RECURSW) \
    X("norecurse", 0, NRECRSW) \
    X("total", 0, TOTALSW) \
    X("nototal", 0, NTOTLSW) \
    X("list", 0, LISTSW) \
    X("nolist", 0, NLISTSW) \
    X("print", 0, PRNTSW) \
    X("noprint", 0, NPRNTSW) \
    X("push", 0, PUSHSW) \
    X("pop", 0, POPSW) \
    X("version", 0, VERSIONSW) \
    X("help", 0, HELPSW) \

#define X(sw, minchars, id) id,
DEFINE_SWITCH_ENUM(FOLDER);
#undef X

#define X(sw, minchars, id) { sw, minchars, id },
DEFINE_SWITCH_ARRAY(FOLDER, switches);
#undef X

static bool fshort;	        /* output only folder names                 */
static int fcreat   = 0;	/* should we ask to create new folders?     */
static bool fpack;      	/* are we packing the folder?               */
static bool fverb;      	/* print actions taken while packing folder */
static int fheader  = 0;	/* should we output a header?               */
static bool frecurse;   	/* recurse through subfolders               */
static int ftotal   = 0;	/* should we output the totals?             */
static bool all;        	/* should we output all folders             */

static int total_folders = 0;	/* total number of folders                  */

static char *nmhdir;
static char *stack = "Folder-Stack";
static char folder[BUFSIZ];

/*
 * Structure to hold information about
 * folders as we scan them.
 */
struct FolderInfo {
    char *name;
    int nummsg;
    int curmsg;
    int lowmsg;
    int hghmsg;
    int others;		/* others == 1 if other files in folder */
    int error;		/* error == 1 for unreadable folder     */
};

/*
 * Dynamically allocated space to hold
 * all the folder information.
 */
static struct FolderInfo *fi;
static int maxFolderInfo;

/*
 * static prototypes
 */
static int get_folder_info (char *, char *);
static crawl_callback_t get_folder_info_callback;
static void print_folders (void);
static int sfold (struct msgs *, char *);
static void readonly_folders (void);

/*
 * Function for printing error message if folder does not exist with
 * -nocreate.
 */
static void
nonexistent_folder (int status)
{
    NMH_UNUSED (status);
    die("folder %s does not exist", folder);
}


int
main (int argc, char **argv)
{
    int printsw = -1;
    bool listsw = false;
    bool pushsw = false;
    bool popsw = false;
    char *cp, *dp, *msg = NULL, *argfolder = NULL;
    char **ap, **argp, buf[BUFSIZ], **arguments;

    if (nmh_init(argv[0], true, true)) { return 1; }

    /*
     * If program was invoked with name ending
     * in `s', then add switch `-all'.
     */
    all = has_suffix_c(argv[0], 's');

    arguments = getarguments (invo_name, argc, argv, 1);
    argp = arguments;

    while ((cp = *argp++)) {
	if (*cp == '-') {
	    switch (smatch (++cp, switches)) {
		case AMBIGSW: 
		    ambigsw (cp, switches);
		    done (1);
		case UNKWNSW: 
		    die("-%s unknown", cp);

		case HELPSW: 
		    snprintf (buf, sizeof(buf), "%s [+folder] [msg] [switches]",
			invo_name);
		    print_help (buf, switches, 1);
		    done (0);
		case VERSIONSW:
		    print_version(invo_name);
		    done (0);

		case ALLSW: 
		    all = true;
		    continue;

		case NALLSW:
		    all = false;
		    continue;

		case CREATSW: 
		    fcreat = 1;
		    continue;
		case NCREATSW: 
		    fcreat = -1;
		    continue;

		case FASTSW: 
		    fshort = true;
		    continue;
		case NFASTSW: 
		    fshort = false;
		    continue;

		case HDRSW: 
		    fheader = 1;
		    continue;
		case NHDRSW: 
		    fheader = -1;
		    continue;

		case PACKSW: 
		    fpack = true;
		    continue;
		case NPACKSW: 
		    fpack = false;
		    continue;

		case VERBSW:
		    fverb = true;
		    continue;
		case NVERBSW:
		    fverb = false;
		    continue;

		case RECURSW: 
		    frecurse = true;
		    continue;
		case NRECRSW: 
		    frecurse = false;
		    continue;

		case TOTALSW: 
		    ftotal = 1;
		    continue;
		case NTOTLSW: 
		    ftotal = -1;
		    continue;

		case PRNTSW: 
		    printsw = 1;
		    continue;
		case NPRNTSW: 
		    printsw = 0;
		    continue;

		case LISTSW: 
		    listsw = true;
		    continue;
		case NLISTSW: 
		    listsw = false;
		    continue;

		case PUSHSW: 
		    pushsw = true;
		    listsw = true;
		    popsw  = false;
		    continue;
		case POPSW: 
		    popsw  = true;
		    listsw = true;
		    pushsw = false;
		    continue;
	    }
	}
	if (*cp == '+' || *cp == '@') {
	    if (argfolder)
		die("only one folder at a time!");
            argfolder = pluspath (cp);
	} else {
	    if (msg)
		die("only one (current) message at a time!");
            msg = cp;
	}
    }

    if (!context_find ("path"))
	free (path ("./", TFOLDER));
    nmhdir = concat (m_maildir (""), "/", NULL);

    /*
     * If not directed via -print/-noprint, we print folder summary
     * info unless if we're working with the folder stack (i.e.,
     * -push, -pop, or -list).
     */
    if (printsw == -1) {
	printsw = !(pushsw || popsw || listsw);
    }

    /* Pushing a folder onto the folder stack */
    if (pushsw) {
	if (!argfolder) {
	    /* If no folder is given, the current folder and */
	    /* the top of the folder stack are swapped.      */
	    if ((cp = context_find (stack))) {
		dp = mh_xstrdup(cp);
		ap = brkstring (dp, " ", "\n");
		argfolder = getcpy(*ap++);
	    } else {
		die("no other folder");
	    }
	    for (cp = mh_xstrdup(getfolder(1)); *ap; ap++)
		cp = add (*ap, add (" ", cp));
	    free (dp);
	    context_replace (stack, cp);	/* update folder stack */
	} else {
	    /* update folder stack */
	    context_replace (stack,
		    (cp = context_find (stack))
		    ? concat (getfolder (1), " ", cp, NULL)
		    : mh_xstrdup(getfolder(1)));
	}
    }

    /* Popping a folder off of the folder stack */
    if (popsw) {
	if (argfolder)
	    die("sorry, no folders allowed with -pop");
	if ((cp = context_find (stack))) {
	    dp = mh_xstrdup(cp);
	    ap = brkstring (dp, " ", "\n");
	    argfolder = getcpy(*ap++);
	} else {
	    die("folder stack empty");
	}
	if (*ap) {
	    /* if there's anything left in the stack */
	    cp = getcpy (*ap++);
	    for (; *ap; ap++)
		cp = add (*ap, add (" ", cp));
	    context_replace (stack, cp);	/* update folder stack */
	} else {
	    context_del (stack);	/* delete folder stack entry from context */
	}
	free (dp);
    }
    if (pushsw || popsw) {
	cp = m_maildir(argfolder);
	if (access (cp, F_OK) == NOTOK)
	    adios (cp, "unable to find folder");
	context_replace (pfolder, argfolder);	/* update current folder   */
	context_save ();		/* save the context file   */
	argfolder = NULL;
    }

    /* Listing the folder stack */
    if (listsw) {
	fputs(argfolder ? argfolder : getfolder (1), stdout);
	if ((cp = context_find (stack))) {
	    dp = mh_xstrdup(cp);
	    for (ap = brkstring (dp, " ", "\n"); *ap; ap++)
		printf (" %s", *ap);
	    free (dp);
	}
	putchar('\n');

	if (!printsw)
	    done (0);
    }

    /* Allocate initial space to record folder information */
    maxFolderInfo = CRAWL_NUMFOLDERS;
    fi = mh_xmalloc (maxFolderInfo * sizeof(*fi));

    /*
     * Scan the folders
     */
    /* change directory to base of nmh directory for crawl_folders */
    if (chdir (nmhdir) == NOTOK)
	adios (nmhdir, "unable to change directory to");
    if (all || ftotal > 0) {
	/*
	 * If no folder is given, do them all
	 */
	if (!argfolder) {
	    if (msg)
		inform("no folder given for message %s, continuing...", msg);
	    readonly_folders (); /* do any readonly folders */
	    cp = context_find(pfolder);
	    strncpy (folder, FENDNULL(cp), sizeof(folder));
	    crawl_folders (".", get_folder_info_callback, NULL);
	} else {
	    strncpy (folder, argfolder, sizeof(folder));
	    if (get_folder_info (argfolder, msg)) {
		context_replace (pfolder, argfolder);/* update current folder */
		context_save ();		     /* save the context file */
	    }
	    /*
	     * Since recurse wasn't done in get_folder_info(),
	     * we still need to list all level-1 sub-folders.
	     */
	    if (!frecurse)
		crawl_folders (folder, get_folder_info_callback, NULL);
	}
    } else {
	strncpy (folder, argfolder ? argfolder : getfolder (1), sizeof(folder));

	/*
	 * Check if folder exists.  If not, then see if
	 * we should create it, or just exit.
	 */
        create_folder (m_maildir (folder), fcreat, nonexistent_folder);

	if (get_folder_info (folder, msg) && argfolder) {
	    /* update current folder */
	    context_replace (pfolder, argfolder);
	    }
    }

    /*
     * Print out folder information
     */
    if (printsw)
	print_folders();

    context_save ();	/* save the context file */
    done (0);
    return 1;
}

static int
get_folder_info_body (char *fold, char *msg, bool *crawl_children)
{
    int	i, retval = 1;
    struct msgs *mp = NULL;

    i = total_folders++;

    /*
     * if necessary, reallocate the space
     * for folder information
     */
    if (total_folders >= maxFolderInfo) {
	maxFolderInfo += CRAWL_NUMFOLDERS;
	fi = mh_xrealloc (fi, maxFolderInfo * sizeof(*fi));
    }

    fi[i].name   = fold;
    fi[i].nummsg = 0;
    fi[i].curmsg = 0;
    fi[i].lowmsg = 0;
    fi[i].hghmsg = 0;
    fi[i].others = 0;
    fi[i].error  = 0;

    if ((ftotal > 0) || !fshort || msg || fpack) {
	/*
	 * create message structure and get folder info
	 */
	if (!(mp = folder_read (fold, fpack))) {
	    inform("unable to read folder %s, continuing...", fold);
	    *crawl_children = false;
	    return 0;
	}

	/* set the current message */
	if (msg && !sfold (mp, msg))
	    retval = 0;

	if (fpack) {
	    if (folder_pack (&mp, fverb) == -1) {
		*crawl_children = false; /* to please clang static analyzer */
		done (1);
	    }
	    seq_save (mp);		/* synchronize the sequences */
	    context_save ();	/* save the context file     */
	}

	/* record info for this folder */
	if ((ftotal > 0) || !fshort) {
	    fi[i].nummsg = mp->nummsg;
	    fi[i].curmsg = mp->curmsg;
	    fi[i].lowmsg = mp->lowmsg;
	    fi[i].hghmsg = mp->hghmsg;
	    fi[i].others = other_files (mp);
	}

	folder_free (mp); /* free folder/message structure */
    }

    *crawl_children = (frecurse && (fshort || fi[i].others)
		       && (fi[i].error == 0));
    return retval;
}

static bool
get_folder_info_callback (char *fold, void *baton)
{
    bool crawl_children;
    NMH_UNUSED (baton);

    get_folder_info_body (fold, NULL, &crawl_children);
    fflush (stdout);
    return crawl_children;
}

static int
get_folder_info (char *fold, char *msg)
{
    bool crawl_children;
    int retval;

    retval = get_folder_info_body (fold, msg, &crawl_children);

    if (crawl_children) {
	crawl_folders (fold, get_folder_info_callback, NULL);
    }

    return retval;
}

/*
 * Print folder information
 */

static void
print_folders (void)
{
    int i, len;
    bool hasempty = false;
    bool curprinted;
    int maxlen = 0, maxnummsg = 0, maxlowmsg = 0;
    int maxhghmsg = 0, maxcurmsg = 0, total_msgs = 0;
    int nummsgdigits, lowmsgdigits;
    int hghmsgdigits, curmsgdigits;
    char tmpname[BUFSIZ];

    /*
     * compute a few values needed to for
     * printing various fields
     */
    for (i = 0; i < total_folders; i++) {
	/* length of folder name */
	len = strlen (fi[i].name);
	if (len > maxlen)
	    maxlen = len;

	/* If folder has error, skip the rest */
	if (fi[i].error)
	    continue;

	/* calculate total number of messages */
	total_msgs += fi[i].nummsg;

	/* maximum number of messages */
	if (fi[i].nummsg > maxnummsg)
	    maxnummsg = fi[i].nummsg;

	/* maximum low message */
	if (fi[i].lowmsg > maxlowmsg)
	    maxlowmsg = fi[i].lowmsg;

	/* maximum high message */
	if (fi[i].hghmsg > maxhghmsg)
	    maxhghmsg = fi[i].hghmsg;

	/* maximum current message */
	if (fi[i].curmsg >= fi[i].lowmsg &&
	    fi[i].curmsg <= fi[i].hghmsg &&
	    fi[i].curmsg > maxcurmsg)
	    maxcurmsg = fi[i].curmsg;

	/* check for empty folders */
	if (fi[i].nummsg == 0)
	    hasempty = true;
    }
    nummsgdigits = num_digits (maxnummsg);
    lowmsgdigits = num_digits (maxlowmsg);
    hghmsgdigits = num_digits (maxhghmsg);
    curmsgdigits = num_digits (maxcurmsg);

    if (hasempty && nummsgdigits < 2)
	nummsgdigits = 2;

    /*
     * Print the header
     */
    if (fheader > 0 || (all && !fshort && fheader >= 0))
	printf ("%-*s %*s %-*s; %-*s %*s\n",
		maxlen+1, "FOLDER",
		nummsgdigits + 13, "# MESSAGES",
		lowmsgdigits + hghmsgdigits + 4, " RANGE",
		curmsgdigits + 4, "CUR",
		9, "(OTHERS)");

    /*
     * Print folder information
     */
    if (all || fshort || ftotal < 1) {
	for (i = 0; i < total_folders; i++) {
	    if (fshort) {
		puts(fi[i].name);
		continue;
	    }

	    /* Add `+' to end of name, if folder is current */
	    if (strcmp (folder, fi[i].name))
		snprintf (tmpname, sizeof(tmpname), "%s", fi[i].name);
	    else
		snprintf (tmpname, sizeof(tmpname), "%s+", fi[i].name);

	    if (fi[i].error) {
		printf ("%-*s is unreadable\n", maxlen+1, tmpname);
		continue;
	    }

	    printf ("%-*s ", maxlen+1, tmpname);

	    curprinted = false; /* remember if we print cur */
	    if (fi[i].nummsg == 0) {
		printf ("has %*s messages%*s",
			nummsgdigits, "no",
			fi[i].others ? lowmsgdigits + hghmsgdigits + 5 : 0, "");
	    } else {
		printf ("has %*d message%1s  (%*d-%*d)",
			nummsgdigits, fi[i].nummsg,
			PLURALS(fi[i].nummsg),
			lowmsgdigits, fi[i].lowmsg,
			hghmsgdigits, fi[i].hghmsg);
		if (fi[i].curmsg >= fi[i].lowmsg && fi[i].curmsg <= fi[i].hghmsg) {
		    curprinted = true;
		    printf ("; cur=%*d", curmsgdigits, fi[i].curmsg);
		}
	    }

	    if (fi[i].others)
		printf (";%*s (others)", curprinted ? 0 : curmsgdigits + 6, "");
	    puts(".");
	}
    }

    /*
     * Print folder/message totals
     */
    if (ftotal > 0 || (all && !fshort && ftotal >= 0)) {
	if (all)
	    putchar('\n');
	printf ("TOTAL = %d message%s in %d folder%s.\n",
		total_msgs, PLURALS(total_msgs),
		total_folders, PLURALS(total_folders));
    }

    fflush (stdout);
}

/*
 * Set the current message and synchronize sequences
 */

static int
sfold (struct msgs *mp, char *msg)
{
    /* parse the message range/sequence/name and set SELECTED */
    if (!m_convert (mp, msg))
	return 0;

    if (mp->numsel > 1) {
	inform("only one message at a time!, continuing...");
	return 0;
    }
    seq_setprev (mp);		/* set the previous-sequence     */
    seq_setcur (mp, mp->lowsel);/* set current message           */
    seq_save (mp);		/* synchronize message sequences */
    context_save ();		/* save the context file         */

    return 1;
}


/*
 * Do the read only folders
 */

static void
readonly_folders (void)
{
    int	atrlen;
    char atrcur[BUFSIZ];
    struct node *np;

    snprintf (atrcur, sizeof(atrcur), "atr-%s-", current);
    atrlen = strlen (atrcur);

    for (np = m_defs; np; np = np->n_next)
	if (ssequal (atrcur, np->n_name)
		&& !ssequal (nmhdir, np->n_name + atrlen))
	    get_folder_info (np->n_name + atrlen, NULL);
}
